include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,382 @@
+			const ICPUImage::SCreationParams& outParams = outImg->getCreationParameters();
+			const auto inFormat = inParams.format;
+			const auto outFormat = outParams.format;
+			const auto inBlockDims = asset::getBlockDimensions(inFormat);
+			const auto outBlockDims = asset::getBlockDimensions(outFormat);
+			const auto texelByteSize = asset::getTexelOrBlockBytesize(outFormat);
+			const auto* const inData = reinterpret_cast<const uint8_t*>(inImg->getBuffer()->getPointer());
+			auto* const outData = reinterpret_cast<uint8_t*>(outImg->getBuffer()->getPointer());
+			const core::SRange<const IImage::SBufferCopy> outRegions = outImg->getRegions(state->outMipLevel);
+
+			const auto inMipLevel = state->inMipLevel;
+			const auto outMipLevel = state->outMipLevel;
+			const auto inBaseLayer = state->inBaseLayer;
+			const auto outBaseLayer = state->outBaseLayer;
+			const auto layerCount = state->inLayerCount;
+			assert(layerCount == state->outLayerCount); // validation bug?
+
+			const auto inOffset = state->inOffset;
+			const auto outOffset = state->outOffset;
+			const auto inExtent = state->inExtent;
+			const auto outExtent = state->outExtent;
+
+			const auto inOffsetBaseLayer = state->inOffsetBaseLayer;
+			const auto outOffsetBaseLayer = state->outOffsetBaseLayer;
+			const auto inExtentLayerCount = state->inExtentLayerCount;
+			const auto outExtentLayerCount = state->outExtentLayerCount;
+			const auto inLimit = inOffsetBaseLayer + inExtentLayerCount;
+			const auto outLimit = outOffsetBaseLayer + outExtentLayerCount;
+
+			CMatchedSizeInOutImageFilterCommon::CommonExecuteData commonExecuteData =
+			{
+				inImg,
+				outImg,
+				inParams,
+				outParams,
+				inParams.format,
+				outParams.format,
+				getTexelOrBlockBytesize(inParams.format),
+				getTexelOrBlockBytesize(outParams.format),
+				reinterpret_cast<const uint8_t*>(inImg->getBuffer()->getPointer()),
+				reinterpret_cast<uint8_t*>(outImg->getBuffer()->getPointer()),
+				inImg->getRegions(state->inMipLevel),
+				outRegions,
+				outRegions.begin(), {}, {}
+			};
+
+			auto allocateCache = [](core::smart_refctd_ptr<asset::ICPUBuffer> cache, size_t size)
+			{
+				auto newPointer = reinterpret_cast<uint8_t*>(_IRR_ALIGNED_MALLOC(size, _IRR_SIMD_ALIGNMENT));
+				cache = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(size, newPointer, core::adopt_memory);
+			};
+
+			allocateCache(state->rowSumCache, outExtent.X * texelByteSize);
+			allocateCache(state->columnSumCache, outExtent.Y * texelByteSize);
+
+			auto perOutputRegion = [state]( const CMatchedSizeInOutImageFilterCommon::CommonExecuteData& commonExecuteData, CBasicImageFilterCommon::clip_region_functor_t& clip) -> bool
+			{
+				const auto blockDims = asset::getBlockDimensions(inFormat);
+
+				auto sum = [&commonExecuteData, &blockDims](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
```

Comment:
first you want to clear your scratch memory to 0s (`memset(,0,)` or `std::fill`)
then you want to decode the input image into scratch via the `executePerRegion`
then you want to peacefully do your own thing with the scratch (the sum)
then you want to save the sum from scratch to output image via the `executePerRegion`

Reply:
Adjusted